### PR TITLE
1:1 mapping between events and throttle methods

### PR DESCRIPTION
**Pull request by sparrowj**
Throttle wasn't actually working in previous versions, because we were using anonymous functions as the method to throttle, so backburner would never get a handle to the throttled method.

### CODE REVIEW

File: addon/services/unified-event-handler.js [hunk 1]
```diff
@@ -106,7 +106,7 @@ export default Ember.Service.extend(Ember.Evented, {
     if (!handlerInfo) {
       // Add new DOM event listener since there is none
       let emberEventName = `${eventName}.${generateId()}`;
-      let trigger = this.triggerEvent.bind(this, emberEventName);
+      let trigger = this.triggerEvent.bind(this, target, eventName);
```
Comment (rwjblue) [inline addon/services/unified-event-handler.js]: Why is this change needed? Also, It seems like `emberEventName` isn't used here any more.

File: addon/services/unified-event-handler.js [hunk 2]
```diff
@@ -120,6 +120,7 @@ export default Ember.Service.extend(Ember.Evented, {
         emberEventName,
         targetElement,
         emberHandlers: [],
+        throttledEvent: this._generateThrottledTriggerFn(),
```
Comment (rwjblue) [inline addon/services/unified-event-handler.js]: In general, I'd like to see this call `throttledEventCallback` or something that indicates its a function...

File: addon/services/unified-event-handler.js [hunk 3]
```diff
@@ -217,18 +218,37 @@ export default Ember.Service.extend(Ember.Evented, {
...
   },
 
+  /**
+   * Creates a function that is used as a handler when throttling events. This makes it possible
+   * for the same throttler function to be used for the same event name
+   * @private
+   * @return {function}
+   */
+  _generateThrottledTriggerFn() {
+    return function({ target, eventName } = {}) {
```
Comment (rwjblue) [inline addon/services/unified-event-handler.js]: should probably return an arrow function since we require valid `this` anyways (then we can tweak the `run.throttle` to avoid re-binding...

File: addon/services/unified-event-handler.js (hunk 1 above)
Comment (sparrowj) [reply addon/services/unified-event-handler.js]: This change is needed because I need `eventName` to lookup `handlerInfo` later on. FYI, `eventName` would be `scroll`, whereas `emerEventName` would be `scroll.1`, or `scroll.2`, etc.

File: addon/services/unified-event-handler.js (hunk 2 above)
Comment (sparrowj) [reply addon/services/unified-event-handler.js]: Agree, will change.

File: addon/services/unified-event-handler.js (hunk 3 above)
Comment (sparrowj) [reply addon/services/unified-event-handler.js]: Spoke offline. Throttle can only be called with a target (target, fn). Going to leave as-is

File: addon/services/unified-event-handler.js [hunk 4]
```diff
@@ -106,7 +106,8 @@ export default Ember.Service.extend(Ember.Evented, {
     if (!handlerInfo) {
       // Add new DOM event listener since there is none
       let emberEventName = `${eventName}.${generateId()}`;
-      let trigger = this.triggerEvent.bind(this, emberEventName);
+      const throttledEventCallback = this._generateEventThrottler(emberEventName);
```
Comment (stefanpenner) [inline addon/services/unified-event-handler.js]: this being a function seems uneeded, why not just inline `() => this.trigger(emberEventName);` ?

File: addon/services/unified-event-handler.js [hunk 5]
```diff
@@ -217,18 +219,25 @@ export default Ember.Service.extend(Ember.Evented, {
...
    * @return {Void}
    */
-  triggerEvent(eventName) {
-    const throttleId = Ember.run.throttle(this, () => {
-      let index = this._throttledEventTimers.indexOf(throttleId);
-      this._throttledEventTimers.splice(index, 1);
-      this.trigger(eventName);
-    }, EVENT_INTERVAL);
-
+  _runThrottler(throttledEventCallback) {
```
Comment (stefanpenner) [inline addon/services/unified-event-handler.js]: `_runThrottler` reads strangely, `_runThrottle` might be better

File: tests/unit/services/unified-event-handler-test.js [hunk 6]
```diff
@@ -227,9 +232,11 @@ test('triggerEvent triggers the event at a throttled rate', function(assert) {
   let throttleSpy = sandbox.spy(Ember.run, 'throttle');
```
Comment (stefanpenner) [inline tests/unit/services/unified-event-handler-test.js]: do we care if throttle is called? This feels needless coupling. Ultimately, all we care is that the callbacks1 and 2 are called appropriately.  We should test that, and only that.

File: tests/unit/services/unified-event-handler-test.js [hunk 7]
```diff
@@ -205,7 +210,7 @@ test('unregister destroys the DOM handler for an event after all callbacks have
 
   let callback1Spy = sandbox.stub();
```
Comment (stefanpenner) [inline tests/unit/services/unified-event-handler-test.js]: no need to suffix this with extra words. `callback1` and `callback2` are sufficient (but because over specificity is not required and because in-this case they are not spies).

File: tests/unit/services/unified-event-handler-test.js [hunk 8]
```diff
@@ -94,7 +99,7 @@ test('register binds multiple callbacks to the event on the specified target but
 
   let callback1Stub = sandbox.stub();
   let callback2Stub = sandbox.stub();
-  let triggerSpy = sandbox.spy(service, 'triggerEvent');
+  let triggerSpy = sandbox.spy(service, '_runThrottler');
```
Comment (stefanpenner) [inline tests/unit/services/unified-event-handler-test.js]: I'm not sure spying on this does anything more then couple the tests to the internal implementation. 

As long as the callbacks are invoked appropriately the test is doing its job, any extra stuff is merely tech debt.

File: tests/unit/services/unified-event-handler-test.js [hunk 9]
```diff
@@ -45,7 +45,12 @@ test('unregisters event listeners when service is destroyed', function(assert) {
 test('cancels throttled events when service is destroyed', function(assert) {
   assert.expect(1);
 
-  service.triggerEvent('scroll');
+  let callbackStub = sandbox.stub();
```
Comment (stefanpenner) [inline tests/unit/services/unified-event-handler-test.js]: `callback`

File: tests/unit/services/unified-event-handler-test.js (hunk 8 above)
Comment (sparrowj) [reply tests/unit/services/unified-event-handler-test.js]: Yeah, I agree. No point in spying on an internal method, esp. since there is essentially no logic between the tested method and the spied method.

File: addon/services/unified-event-handler.js (hunk 5 above)
Comment (sparrowj) [reply addon/services/unified-event-handler.js]: Fixed.

File: addon/services/unified-event-handler.js (hunk 4 above)
Comment (sparrowj) [reply addon/services/unified-event-handler.js]: Good call.

File: tests/unit/services/unified-event-handler-test.js (hunk 7 above)
Comment (sparrowj) [reply tests/unit/services/unified-event-handler-test.js]: Fixed here, and a couple of other places. For these PR's, I figured it's better to touch as little existing code as I can, to keep the PR cleaner. Happy to come back in and do some spring cleaning if you think that's necessary.

File: tests/unit/services/unified-event-handler-test.js (hunk 6 above)
Comment (sparrowj) [reply tests/unit/services/unified-event-handler-test.js]: Thought about this for a while. While I think the fact that methods are throttled is important, there is no logic wrapping the throttling functionality, so probably it's not necessary to unit tests. Will remove.

**Comment by stefanpenner** [comment]
released as v1.0.4 🎉
